lib/hashset.rs

```diff
@@ -0,0 +1,387 @@
+// `im::HashSet::iter` returns the non-deterministic iterator (which can
+// mess up DDlog in ways one can easily miss unless they test on big
+// enough sets).
+
+// impl<T> Deref for HashSet<T> {
+//     type Target = IMHashSet<T>;
+//
+//     fn deref(//self) -> //Self::Target {
+//         //self.set
+//     }
+// }
+//
+// impl<T> DerefMut for HashSet<T> {
+//     fn deref_mut(//mut self) -> //mut Self::Target {
+//         //mut self.set
+//     }
+// }
+
+impl<T: FromRecord + Hash + Eq + Clone + Ord> FromRecord for HashSet<T> {
+    fn from_record(val: &Record) -> StdResult<Self, String> {
+        match val {
+            Record::Array(_, args) => StdResult::from_iter(args.iter().map(T::from_record)),
+            v => T::from_record(v).map(Self::unit),
+        }
+    }
+}
+
+impl<T: IntoRecord + Hash + Eq + Clone + Ord> IntoRecord for HashSet<T> {
+    fn into_record(self) -> Record {
+        Record::Array(
+            CollectionKind::Set,
+            self.into_iter().map(|x| x.into_record()).collect(),
+        )
+    }
+}
+
+// Set update semantics: update contains values that are in one of the sets
+// but not the other.
+impl<T: FromRecord + Hash + Eq + Clone + Ord> Mutator<HashSet<T>> for Record {
+    fn mutate(&self, set: &mut HashSet<T>) -> StdResult<(), String> {
+        let upd = <HashSet<T>>::from_record(self)?;
+        for v in upd.into_iter() {
+            if set.remove(&v).is_none() {
+                set.insert(v);
+            }
+        }
+        Ok(())
+    }
+}
+
+impl<T: Hash + Eq + Debug + Ord> Debug for HashSet<T> {
+    fn fmt(&self, f: &mut Formatter) -> FmtResult {
+        f.debug_set().entries(self.iter()).finish()
+    }
+}
+
+pub fn hashset_size<T>(s: &HashSet<T>) -> std_usize {
+    s.len() as std_usize
+}
+
+pub fn hashset_empty<T>() -> HashSet<T> {
+    HashSet::new()
+}
+
+pub fn hashset_singleton<T: Hash + Eq + Clone>(v: &T) -> HashSet<T> {
+    HashSet::unit((*v).clone())
+}
+
+pub fn hashset_insert<T: Hash + Eq + Clone>(s: &mut HashSet<T>, v: &T) {
+    s.insert(v.clone());
+}
+
+pub fn hashset_insert_imm<T: Hash + Eq + Clone>(s: &HashSet<T>, v: &T) -> HashSet<T> {
+    s.update((*v).clone())
+}
+
+pub fn hashset_remove<T: Hash + Eq + Clone>(s: &mut HashSet<T>, v: &T) {
+    s.remove(v);
+}
+
+pub fn hashset_remove_imm<T: Hash + Eq + Clone>(s: &HashSet<T>, v: &T) -> HashSet<T> {
+    s.without(v)
+}
+
+pub fn hashset_contains<T: Hash + Eq>(s: &HashSet<T>, v: &T) -> bool {
+    s.contains(v)
+}
+
+pub fn hashset_is_empty<T>(s: &HashSet<T>) -> bool {
+    s.is_empty()
+}
+
+pub fn hashset_nth<T: Hash + Clone + Ord>(s: &HashSet<T>, n: &std_usize) -> DDOption<T> {
+    option2std(s.iter().nth(*n as usize).cloned())
+}
+
+pub fn hashset_to_vec<T: Hash + Clone + Ord>(s: &HashSet<T>) -> DDVec<T> {
+    DDVec {
+        vec: s.iter().cloned().collect(),
+    }
+}
+
+pub fn hashset_union<T: Hash + Eq + Clone>(s1: &HashSet<T>, s2: &HashSet<T>) -> HashSet<T> {
+    s1.clone().union(s2.clone())
```

Comment:
Why clone?

Reply:
The `union` function consumes both its arguments, so we cannot pass them by reference.